client/components/hoc/protect-form/index.js

```diff
@@ -0,0 +1,81 @@
+/**
+ * External dependencies
+ */
+import React, { Component } from 'react';
+import debugModule from 'debug';
+import page from 'page';
+import i18n from 'i18n-calypso';
+import { without } from 'lodash';
+
+/**
+ * Module variables
+ */
+const debug = debugModule( 'calypso:protect-form' );
+const confirmText = i18n.translate( 'You have unsaved changes. Are you sure you want to leave this page?' );
+const beforeUnloadText = i18n.translate( 'You have unsaved changes.' );
+let formsChanged = [];
+
+export const protectForm = WrappedComponent => {
+	class ProtectedFormComponent extends Component {
+		componentDidMount() {
+			window.addEventListener( 'beforeunload', this.warnIfChanged );
+		}
+
+		componentWillUnmount() {
+			window.removeEventListener( 'beforeunload', this.warnIfChanged );
+			formsChanged = without( formsChanged, this );
+		}
+
+		warnIfChanged = event => {
+			if ( ! formsChanged.length ) {
+				return;
+			}
+			debug( 'unsaved form changes detected' );
+			( event || window.event ).returnValue = beforeUnloadText;
+			return beforeUnloadText;
+		};
+
+		markChanged = () => {
+			if ( -1 === formsChanged.indexOf( this ) ) {
+				formsChanged.push( this );
+			}
+		};
+
+		markSaved = () => {
+			const index = formsChanged.indexOf( this );
+			if ( -1 === index ) {
+				return;
+			}
+			formsChanged.splice( index, 1 );
+		}
+
+		render() {
+			return (
+				<WrappedComponent
+					markChanged={ this.markChanged }
+					markSaved={ this.markSaved }
+					{ ...this.props }
+				/>
+			);
+		}
+	}
+
+	return ProtectedFormComponent;
+};
+
+export const checkFormHandler = ( context, next ) => {
```

Comment:
`page.js` has come up recently as one of the sources of performance issues we're seeing, given the number of middlewares that are run on every page transition. Do you think there might be a way to avoid this? My thinking is perhaps a Redux middleware that checks on dispatch of a `ROUTE_SET` action?